### PR TITLE
[swiftc] Fixed for Cygwin

### PR DESCRIPTION
**Pull request by tinysun212**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

Fixed for the difference of Cygwin with other Windows variants (MSVC,
Itanium, MinGW).
- The platform name is renamed to "cygwin" from "windows" which is used
  for searching the standard libraries.
- The consideration for DLL storage class (DllExport/DllImport) is not
  required for Cygwin and MinGW. There is no problem when linking in
  these environment.
- Cygwin should use large memory model as default.(This may be changed
  if someone ports to 32bit)
- Cygwin and MinGW should use the autolink feature in the sameway of
  Linux due to the linker's limit.

### CODE REVIEW

File: tools/driver/CMakeLists.txt [hunk 1]
```diff
@@ -38,6 +38,11 @@ if(SWIFT_HOST_VARIANT MATCHES "${SWIFT_DARWIN_VARIANTS}")
       LINK_FLAGS " -lobjc -Wl,-force_load,\"${SWIFT_REPL_ARCLITE}\"")
 endif()
 
+if("${CMAKE_SYSTEM_NAME}" STREQUAL "CYGWIN")
+  set_property(TARGET swift APPEND_STRING PROPERTY
+      LINK_FLAGS " -Wl,--allow-multiple-definition")
```
Comment (gribozavr) [inline tools/driver/CMakeLists.txt]: Why is this change necessary?

Comment (tinysun212) [inline tools/driver/CMakeLists.txt]: Linker generates the error of multiple definition of `guard variable for swift::ImmutablePointerSetFactory<swift::SILInstruction>::EmptyPtrSet` in `libswiftSILOptimizer.a`.
And the `EmptyPtrSet` is defined at `include/swift/Basic/ImmutablePointerSet.h`.

The symbol is defined in the object files compiled from all `lib/SILOptimizer/ARC/*.cpp` (10 files).  I think it is related to 'template instantiation'. But I don't know whether those sources are good or bad.

I added this flag instead of touching sources.

Comment (compnerd) [inline tools/driver/CMakeLists.txt]: The GNU ld linker has a very poor approximation of COFF support.  It fails constantly under a myriad of conditions, including COMDAT symbols.  The `--allow-multiple-definitions` is attempting to compensate for a poor implementation of a COFF linker.  To make things worse, this is something which is hardcoded into the gcc specs IIRC, and has even partially made its way into clang :-(.  I for one welcome our new LLVM overlords^W^W^W^Wlld.

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,6 +4,10 @@ if(POLICY CMP0051)
   cmake_policy(SET CMP0051 NEW)
 endif()
 
+if(POLICY CMP0054)
```
Comment (jrose-apple) [inline CMakeLists.txt]: >Only interpret if() arguments as variables or keywords when unquoted.

Why do we need this to be turned off? Can't we just quote things correctly?

File: lib/IRGen/IRGenModule.cpp [hunk 3]
```diff
@@ -980,7 +983,7 @@ void IRGenModule::emitAutolinkInfo() {
                                  llvm::GlobalValue::CommonLinkage,
                                  llvm::Constant::getNullValue(Int1Ty),
                                  buf.str());
-    if (Triple.isOSBinFormatCOFF())
+    if (Triple.isOSBinFormatCOFF() && !Triple.isOSCygMing())
```
Comment (jrose-apple) [inline lib/IRGen/IRGenModule.cpp]: This is repeated so much I think we should just have a predicate for it on IRGenModule.

File: lib/Markup/Markup.cpp [hunk 4]
```diff
@@ -15,6 +15,9 @@
 #include "swift/AST/Comment.h"
 #include "swift/Markup/LineList.h"
 #include "swift/Markup/Markup.h"
+#if defined(__CYGWIN__)
```
Comment (jrose-apple) [inline lib/Markup/Markup.cpp]: We have our own fork of CMark; please submit this as a patch there (or better yet submit it upstream).

**Comment by jrose-apple** [comment]
@compnerd Mind looking this over as well?

File: CMakeLists.txt (hunk 2 above)
Comment (tinysun212) [inline CMakeLists.txt]: I found that `stdlib/public/core/CMakeLists.txt` uses `if(CMAKE_SYSTEM_NAME STREQUAL ...` comparing with "Darwin" and "FreeBSD". To be turned on, we should scan all CMakeLists.txt and *.cmake files and test build for all platforms. That was not simple job to me.

Comment (jrose-apple) [inline CMakeLists.txt]: That looks correct to me. CMP0054 controls `if("CMAKE_SYSTEM_NAME" STREQUAL ...`.

Comment (compnerd) [inline CMakeLists.txt]: Im not sure I understand this.  Why would this change suddenly become a problem?  We have been using the new policy, so it would have broken if it wasn't compatible atm, right?

File: lib/Basic/Platform.cpp [hunk 5]
```diff
@@ -118,6 +118,12 @@ StringRef swift::getPlatformNameForTriple(const llvm::Triple &triple) {
   case llvm::Triple::FreeBSD:
     return "freebsd";
   case llvm::Triple::Win32:
+    if (triple.isWindowsCygwinEnvironment())
+      return "cygwin";
+
+    if (triple.isWindowsGNUEnvironment())
+      return "mingw";
+
```
Comment (compnerd) [inline lib/Basic/Platform.cpp]: I think it would be nicer to use a switch here instead.  That way if we run into similar issues in the future, its easier to extend.

File: lib/IRGen/GenDecl.cpp [hunk 6]
```diff
@@ -937,7 +937,7 @@ void IRGenModule::emitVTableStubs() {
     // For each eliminated method symbol create an alias to the stub.
     auto *alias = llvm::GlobalAlias::create(llvm::GlobalValue::ExternalLinkage,
                                             F.getName(), stub);
-    if (Triple.isOSBinFormatCOFF())
+    if (Triple.isOSBinFormatCOFF() && !Triple.isOSCygMing())
```
Comment (compnerd) [inline lib/IRGen/GenDecl.cpp]: Does this actually break down on cygwin?

File: lib/IRGen/GenDecl.cpp [hunk 7]
```diff
@@ -1239,7 +1239,8 @@ getIRLinkage(IRGenModule &IGM, SILLinkage linkage, bool isFragile,
 
   const auto ObjFormat = IGM.TargetInfo.OutputObjectFormat;
   bool IsELFObject = ObjFormat == llvm::Triple::ELF;
-  bool IsCOFFObject = ObjFormat == llvm::Triple::COFF;
+  bool UseImportOrExport =
```
Comment (compnerd) [inline lib/IRGen/GenDecl.cpp]: I think that `UseDLLStorage` is a more pithy name.

File: lib/IRGen/IRGen.cpp [hunk 8]
```diff
@@ -490,10 +490,13 @@ static createTargetMachine(IRGenOptions &Opts, ASTContext &Ctx) {
   }
 
   // Create a target machine.
-  llvm::TargetMachine *TargetMachine
-    = Target->createTargetMachine(Triple.str(), CPU,
-                                  targetFeatures, TargetOpts, Reloc::PIC_,
-                                  CodeModel::Default, OptLevel);
+  auto cmodel = CodeModel::Default;
+  if (Triple.isWindowsCygwinEnvironment())
+    cmodel = CodeModel::Large;
```
Comment (compnerd) [inline lib/IRGen/IRGen.cpp]: Why do you need the large code model?

File: lib/IRGen/IRGenModule.cpp [hunk 9]
```diff
@@ -927,9 +929,9 @@ void IRGenModule::emitAutolinkInfo() {
                                        }),
                         AutolinkEntries.end());
 
-  if (TargetInfo.OutputObjectFormat == llvm::Triple::COFF ||
-      TargetInfo.OutputObjectFormat == llvm::Triple::MachO ||
-      Triple.isPS4()) {
+  if ((TargetInfo.OutputObjectFormat == llvm::Triple::COFF &&
+       !Triple.isOSCygMing()) ||
```
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: MinGW can properly handle directives for export.  Why are you disabling this?

File: CMakeLists.txt (hunk 2 above)
Comment (tinysun212) [inline CMakeLists.txt]: I used the version 3.3.2 for cmake which uses 'NEW' policy by default.
Without the policty setting, this cmake version prints many warning messages for almost `if()`  even if the default is 'NEW'.
So I should set the policy to 'OLD' or 'NEW'.

If I set to 'NEW', it will break other platforms which used cmake version 2.x.

File: lib/IRGen/IRGenModule.cpp (hunk 9 above)
Comment (tinysun212) [inline lib/IRGen/IRGenModule.cpp]: MinGW/Cygwin's `ld` recognizes `.drectve ` section for export, but it is limited and is not same to MS  Linker.
It generated the warning "Warning: .drectve OOOO unrecognized" for "-lFoo" option in the section. We can see a part of ld source [here](https://fossies.org/linux/misc/binutils-2.26.1.tar.gz/binutils-2.26.1/ld/deffilep.y). (line 875~883, 934~950)

File: lib/IRGen/IRGen.cpp (hunk 8 above)
Comment (tinysun212) [inline lib/IRGen/IRGen.cpp]: DLL's are loaded above the max address for 32bit. (for example, cygswiftcore.dll is loaded at 0x459F40000).
Without large model, generated code had problems to reference the .data section of DLL's, and generated segment fault.

File: lib/IRGen/GenDecl.cpp (hunk 6 above)
Comment (tinysun212) [inline lib/IRGen/GenDecl.cpp]: dll import did break, but dll export may not break, I simply disabled both for consistency.

File: lib/IRGen/GenDecl.cpp (hunk 7 above)
Comment (tinysun212) [inline lib/IRGen/GenDecl.cpp]: OK. I'll change it.

File: lib/Basic/Platform.cpp (hunk 5 above)
Comment (tinysun212) [inline lib/Basic/Platform.cpp]: Did you mean a nested switch in `case llvm::Triple::Win32:` ?
```
switch (triple.getEnvironment()) {
case llvm::Triple::Cygnus:
  return "cygwin";
case llvm::Triple::GNU:
  return "mingw":
default:
  return "windows";
}
```
I can not imagine how cover those with single switch.

File: CMakeLists.txt (hunk 2 above)
Comment (compnerd) [inline CMakeLists.txt]: LLVM has a minimum requirement for CMake at 3.4 IIRC.  The point is that it should already be NEW for everyone.  Why has no one hit this yet?  No one should be using 2.x.

File: lib/Basic/Platform.cpp (hunk 5 above)
Comment (compnerd) [inline lib/Basic/Platform.cpp]: Yes, a nested switch.  You should be able to do a covered switch there, and have most of the environments be unreachable.

File: lib/IRGen/GenDecl.cpp (hunk 6 above)
Comment (compnerd) [inline lib/IRGen/GenDecl.cpp]: It would be interesting to know how the dllimport breaks.

File: lib/IRGen/IRGen.cpp (hunk 8 above)
Comment (compnerd) [inline lib/IRGen/IRGen.cpp]: Interesting.  Thats something that I somehow did not hit.  However, even for the MSVC environment, we should be using `LARGEADDRESSAWARE` mode.  So, perhaps its best to just always enable the large code model.  I think the selection can just be inlined via a ternary.

File: lib/IRGen/IRGenModule.cpp (hunk 9 above)
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: Im familiar with the directive handling in binutils.  That sounds like something else is going wrong.  I know that the directives generated by LLVM should be properly handled by binutils linker.  We should look at the generated object file and figure out what exactly prevents us from using that infrastructure.  If its a bug in LLVM, we can fix that :-).

File: CMakeLists.txt (hunk 2 above)
Comment (gottesmm) [inline CMakeLists.txt]: We require 3.4.3. I ran into an issue when I changed the requirement. I am nervous about making such a change (since it may change policy defaults) until we do the final branch for swift-3.

When that happens I am going to update the actual requirement.

File: lib/IRGen/IRGenModule.cpp (hunk 3 above)
Comment (tinysun212) [inline lib/IRGen/IRGenModule.cpp]: I'm adding `IRGenModule::useDllStorage()`, many lines will be replaced but not all of them.

File: lib/IRGen/IRGen.cpp (hunk 8 above)
Comment (tinysun212) [inline lib/IRGen/IRGen.cpp]: Sorry, I didn't understand last sentence. 'via a ternary' means all MinGW, Cygwin, MSVC ?

File: lib/IRGen/IRGenModule.cpp (hunk 9 above)
Comment (tinysun212) [inline lib/IRGen/IRGenModule.cpp]: I think LLVM should not emit '-lFoo' into `.drectve` since ld can not recognize it. (in case of Cygwin/MinGW)
Below is the output before patch.
```
$ ./swiftc -c Hello.swift

$ objdump -s -h --section=.drectve Hello.o

Hello.o:     file format pe-x86-64

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  6 .drectve      00000025  0000000000000000  0000000000000000  00000314  2**0
                  CONTENTS, READONLY, DEBUGGING, EXCLUDE
Contents of section .drectve:
 0000 202d6c73 77696674 436f7265 202d6c73   -lswiftCore -ls
 0010 77696674 53776966 744f6e6f 6e655375  wiftSwiftOnoneSu
 0020 70706f72 74                          pport

$ ./swiftc Hello.swift
Warning: .drectve `-lswiftCore ' unrecognized
Warning: corrupt .drectve at end of def file
/tmp/Hello-a57504.o:(.text+0x5f): undefined reference to `_TTSgq5P____TFs27_allocateUninitializedArrayurFBwTGSax_Bp_'
collect2: error: ld returned 1 exit status
clang-3.7: error: linker (via gcc) command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: link command failed with exit code 1 (use -v to see invocation)

$ ./swiftc Hello.swift -lswiftSwiftOnoneSupport
Warning: .drectve `-lswiftCore ' unrecognized
Warning: corrupt .drectve at end of def file

```

File: tools/driver/CMakeLists.txt (hunk 1 above)
Comment (tinysun212) [inline tools/driver/CMakeLists.txt]: I'll remove this part. Updating clang to 3.8.1 with recent released cygwin package, the link flag can be removed.

With some test code, I found the Cygwin clang compiler 3.7.1 generates wrong section type for the data section which has a guard variable for template-static-member. That leads to linker's multiple definition message.

File: CMakeLists.txt (hunk 2 above)
Comment (tinysun212) [inline CMakeLists.txt]: After comparing the CMake in Cygwin and Linux, I realized that I misunderstood the behavior. I'll separate this part to another PR with other approach method.

In CMake 3.4.3, if `CMP0054` policy isn't set to any value, `cmake` uses the OLD behavior and issues a long warning message for every `if()` statement having a quoted variable whose value is an other variable's name. 
For example,
```
set(FOO 1)
set(VAR_A FOO)
set(VAR_B BAR)
if ("${VAR_A}" STREQUAL "FOO")
endif()
if ("${VAR_B}" STREQUAL "BAR")
endif()
```
issues `CMP0054` warning message for `"${VAR_A}"` and not for `"${VAR_B}"`.

Cygwin `cmake` defines the variable `CMAKE_SYSTEM_NAME` as `CYGWIN` and defines the variable `CYGWIN` as 1(TRUE).

Linux `cmake` defines the variable `CMAKE_SYSTEM_NAME` as `Linux` and there is no variable `Linux`.

So, if we didn't set any value to `CMP0054` policy, every `if ("${CMAKE_SYSTEM_NAME}" ...)` generated long warning message in Cygwin, but did not generate in Linux.

Comment (jrose-apple) [inline CMakeLists.txt]: I see. So the right answer is to set it to `NEW`?

Comment (tinysun212) [inline CMakeLists.txt]: Yes, it will be better for Cygwin.
I think other platforms don't have any effect to `CMP0054 NEW`, but it is not proven.
So, I'm preparing a PR in which the policy setting is guarded with Cygwin condition.

File: lib/Markup/Markup.cpp (hunk 4 above)
Comment (tinysun212) [inline lib/Markup/Markup.cpp]: This is not CMark source problem.

There was some problem in dynamic linking with cmark, I simply chose the static link method for a workaround. I dropped this part.

I'll solve the dynamic linking problem later.

**Comment by tinysun212** [comment]
ping?

File: lib/IRGen/IRGen.cpp (hunk 8 above)
Comment (compnerd) [inline lib/IRGen/IRGen.cpp]: Actually, thinking more about it, do you have the code from an instance where this happens?  The code generation issue that you sounds very much like a bug in LLVM.  x86 has fully addressability, so within a 32-bit range should be fine.

File: lib/IRGen/IRGenModule.cpp [hunk 10]
```diff
@@ -446,6 +446,7 @@ llvm::Constant *swift::getRuntimeFn(llvm::Module &Module,
     fn->setCallingConv(cc);
 
     if (llvm::Triple(Module.getTargetTriple()).isOSBinFormatCOFF() &&
+        !llvm::Triple(Module.getTargetTriple()).isOSCygMing() &&
```
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: This is IGM, so you should be able to use `useDLLStorage`.

File: lib/IRGen/IRGenModule.cpp [hunk 11]
```diff
@@ -525,7 +526,8 @@ llvm::Constant *swift::getWrapperFn(llvm::Module &Module,
     auto *globalFnPtr =
         new llvm::GlobalVariable(Module, fnPtrTy, false,
                                  llvm::GlobalValue::ExternalLinkage, 0, symbol);
-    if (llvm::Triple(Module.getTargetTriple()).isOSBinFormatCOFF())
+    if (llvm::Triple(Module.getTargetTriple()).isOSBinFormatCOFF() &&
+        !llvm::Triple(Module.getTargetTriple()).isOSCygMing())
```
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: Same.

File: lib/IRGen/TypeLayoutVerifier.cpp [hunk 12]
```diff
@@ -43,7 +43,7 @@ irgen::emitTypeLayoutVerifier(IRGenFunction &IGF,
                                               /*var arg*/ false);
   auto verifierFn = IGF.IGM.Module.getOrInsertFunction(
       "_swift_debug_verifyTypeLayoutAttribute", verifierFnTy);
-  if (IGF.IGM.Triple.isOSBinFormatCOFF())
+  if (IGF.IGM.Triple.isOSBinFormatCOFF() && !IGF.IGM.Triple.isOSCygMing())
```
Comment (compnerd) [inline lib/IRGen/TypeLayoutVerifier.cpp]: Why not use `IGF.IGM.useDLLStorage()`?

Comment (tinysun212) [inline lib/IRGen/TypeLayoutVerifier.cpp]: I missed it. Updated it just.

File: lib/IRGen/IRGen.cpp (hunk 8 above)
Comment (tinysun212) [inline lib/IRGen/IRGen.cpp]: Now I can show you the sample codes.

This is a line of LLVM of Hello.swift.

`
@_TZvOs11CommandLine5_argcVs5Int32 = external global %Vs5Int32, align 4
`

That is `static Swift.CommandLine._argc : Swift.Int32`,
and is a part of the standard library and declared in data section of `cygswiftCore.dll`,
so the application code should read/write the data section of the dll. 

The generated assembly code is as follows.

In default model,
`movl    %edi, _TZvOs11CommandLine5_argcVs5Int32(%rip)`,

and in large model,
`movabsq $_TZvOs11CommandLine5_argcVs5Int32, %rax`
`movl    %ecx, (%rax)`.

In case of 64bit default memory model, since the DLL is located at higher memory (for example, cygswiftcore.dll is loaded at 0x459F40000 > maximum 32bit), and the generated code only references 32bit address, the executable file didn't work correctly.

If we link the static library instead of dll, it might be work correctly, so I think it is hard to say llvm has a bug.

**Comment by tinysun212** [comment]
Any more reviews ?

File: lib/IRGen/IRGen.cpp (hunk 8 above)
Comment (compnerd) [inline lib/IRGen/IRGen.cpp]: Hmm, doesnt this problem exist beyond just cygwin?  Perhaps it should be a 64-bit windows check here instead?

File: lib/IRGen/IRGenModule.cpp (hunk 9 above)
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: Yeah, we should use `-defaultlib swiftCore.lib` instead.

File: lib/IRGen/IRGen.cpp (hunk 8 above)
Comment (tinysun212) [inline lib/IRGen/IRGen.cpp]: Yes, This will be same in case of MinGW 64bit and Windows MSVC 64bit.
What about 64bit Windows with different arch, such as ARM ?

**Comment by slavapestov** [comment]
Looks like this needs rebase and conflict resolution too

**Comment by jrose-apple** [comment]
This was mostly waiting for @compnerd's sign-off, I think.

File: lib/IRGen/IRGen.cpp (hunk 8 above)
Comment (compnerd) [inline lib/IRGen/IRGen.cpp]: Doesnt 64-bit MSVC also have 2-GB limitations (PE+).

File: lib/IRGen/IRGenModule.cpp [hunk 13]
```diff
@@ -440,6 +440,7 @@ llvm::Constant *swift::getRuntimeFn(llvm::Module &Module,
     fn->setCallingConv(cc);
 
     if (llvm::Triple(Module.getTargetTriple()).isOSBinFormatCOFF() &&
```
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: I think that this should use `useDllStorage()` instead of adding the negation check so that this is more similar to the previous case.

File: lib/IRGen/IRGenModule.cpp [hunk 14]
```diff
@@ -519,7 +520,8 @@ llvm::Constant *swift::getWrapperFn(llvm::Module &Module,
     auto *globalFnPtr = new llvm::GlobalVariable(
         Module, fnPtrTy, false, llvm::GlobalValue::ExternalLinkage, nullptr,
         symbol);
-    if (llvm::Triple(Module.getTargetTriple()).isOSBinFormatCOFF())
+    if (llvm::Triple(Module.getTargetTriple()).isOSBinFormatCOFF() &&
+        !llvm::Triple(Module.getTargetTriple()).isOSCygMing())
```
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: Similar

File: lib/IRGen/IRGenModule.cpp [hunk 15]
```diff
@@ -916,9 +918,9 @@ void IRGenModule::emitAutolinkInfo() {
                                        }),
                         AutolinkEntries.end());
 
-  if (TargetInfo.OutputObjectFormat == llvm::Triple::COFF ||
-      TargetInfo.OutputObjectFormat == llvm::Triple::MachO ||
-      Triple.isPS4()) {
+  if ((TargetInfo.OutputObjectFormat == llvm::Triple::COFF &&
+       (!Triple.isOSCygMing() || Triple.getArch() == llvm::Triple::thumb)) ||
```
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: Why the special case for cygwin on Windows ARM here?

Comment (tinysun212) [inline lib/IRGen/IRGenModule.cpp]: Test script - https://github.com/apple/swift/blob/master/test/IRGen/autolink-coff.swift - asserts the target `thumbv7--windows-gnu` to use the MSVC-style autolink, which uses `.drectve` section.
I created the `autolink-coff-x86.swift` for MinGW (`x86--windows-gnu`) to test the Linux-style autolink.
If we drop/modify the autolink test case for `thumbv7--windows-gnu` , the code will be simpler.

File: lib/IRGen/IRGenModule.cpp (hunk 14 above)
Comment (tinysun212) [inline lib/IRGen/IRGenModule.cpp]: Thanks, I'll change them.

File: lib/IRGen/IRGenModule.cpp (hunk 15 above)
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp]: Lets change the test for that target.  Really, the only Windows ARM target that I know of which is tested/works is Windows itanium (with clang) or msvc (with msvc).

**Comment by tinysun212** [comment]
@compnerd  The test script `autolink-coff.swift` and other reviews were applied.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
Ok, I'm merging this as soon as @cmpnerd gives the go-ahead (or does @compnerd already have commit access)?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by compnerd** [comment]
@slavapestov yeah, I do have commit access :-).
